photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,307 @@
+   */
+  protected def tempColumnName(name: String): String = s"__${name}__${System.currentTimeMillis}__"
+
+  /**
+   * Read a feature tuple from the name, term, and value DataFrame row
+   *
+   * @param row the row from which to read the name, term, value tuple
+   * @return the feature tuple
+   */
+  protected def readFeature = (feature: Row) => {
+    val fieldNames = feature.schema.fieldNames
+
+    if (fieldNames.contains(FieldNames.NAME) && fieldNames.contains(FieldNames.VALUE)) {
+      val name = feature.getAs[String](FieldNames.NAME)
+
+      val term = if (fieldNames.contains(FieldNames.TERM)) {
+        feature.getAs[String](FieldNames.TERM)
+      } else {
+        ""
+      }
+
+      val value = feature.getAs[Number](FieldNames.VALUE).doubleValue
+
+      Some((Utils.getFeatureKey(name, term) -> value))
+    } else {
+      None
+    }
+  }
+
+  /**
+   * A Spark sql udf that merges the features of all the columns into a single column
+   *
+   * @param featureReader the feature reader instance
+   * @param cols Spark sql columns to be merged
+   */
+  protected def mergeFeatures = {
+    val reader = readFeature
+    udf((cols: Seq[Seq[Row]]) =>
+      cols.flatMap { features => features.flatMap(reader(_)) })
+  }
+
+  /**
+   * Drops all the specified columns from the DataFrame
+   *
+   * @param df the source DataFrame
+   * @param cols the column names to drop
+   * @return the transformed DataFrame
+   */
+  protected def dropColumns(df: DataFrame, cols: Seq[String]): DataFrame =
+    cols.foldLeft(df) { case (acc, columnName) => acc.drop(columnName) }
+
+  /**
+   * Merges the feature columns according to the feature column map
+   *
+   * @param df the source DataFrame
+   * @param featureColumnMap a map that specifies how the feature columns should be merged. The keys specify the name
+   *   of the merged destination column, and the values are sets of source columns to merge, e.g.:
+   *
+   *     Map("userFeatures" -> Set("profileFeatures", "titleFeatures"))
+   *
+   *   This configuration merges the "profileFeatures" and "titleFeatures" columns into a single column named
+   *   "userFeatures".
+   * @return the transformed DataFrame
+   */
+  protected def mergeColumns(
+      df: DataFrame,
+      featureColumnMap: Map[String, Set[String]]): DataFrame = {
+
+    val initTempColMappings = Map.empty[String, String]
+
+    // Merge the columns into new temporary columns. We need to create intermediate temporary columns to correctly
+    // handle cases where destination columns are named the same as source columns. Replacing these in-situ could cause
+    // unexpected results when merging subsequent columns.
+    val (merged, tempColMappings) = featureColumnMap.foldLeft((df, initTempColMappings)) {
+      case ((merged, tempColMappings), (destColumn, sourceColumns)) =>
+        val sourceCols = sourceColumns.map(col(_)).toSeq
+        val newCol = mergeFeatures(array(sourceCols:_*))
```

Comment:
It's the second one. Basically the column is materialized according to the udf when we call `merged.withColumn`.